File: scripts/anki-add-card.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

class String
	def into_path
		File::expand_path self
	end
end

class Array
	def pad_to_required
		(self + [''] * 2)[0...6]
	end
end

def cry_with_value how, about_what
	warn "error: #{how} `#{about_what}`"
	exit false
end

def cry about_what
	warn "error: #{about_what}"
	exit false
end

EXISTING_DECKS = %w[Once Freq].freeze
EXISTING_NOTE_TYPES = %w[b d f fb h].freeze

lines = File::open('~/bs/anki-card.html'.into_path).each_line.map(&:strip).to_a

fields = lines.size
cry_with_value 'less than 4 fields', fields if fields < 4
cry_with_value 'more than 6 fields', fields if fields > 6

lines = lines.pad_to_required.to_enum

deck = lines.next
cry_with_value 'invalid deck', deck unless EXISTING_DECKS.include?(deck)
note_type = lines.next
cry_with_value 'invalid note type', note_type unless EXISTING_NOTE_TYPES.include?(note_type)

lines = lines.map do |line|
	line = line.gsub '"', '""'
	'"' + line + '"'
end.to_a

puts lines.join ';'

File: scripts/anki-add-card.rs
// begin Cargo.toml
// [dependencies]
// end Cargo.toml
// /home/axlefublr/r/dot/scripts/anki-add-card.rs
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(dead_code)]

use std::env;
use std::env::args;
use std::error::Error;
use std::fmt;
use std::fs;
use std::fs::File;
use std::fs::OpenOptions;
use std::io::stdin;
use std::io::Read;
use std::io::Write;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;

const VALID_DECKS: [&str; 2] = ["Once", "Freq"];
const VALID_NOTE_TYPES: [&str; 5] = ["b", "d", "f", "fb", "h"];

fn main() -> Result<(), Box<dyn Error>> {
    let contents = fs::read_to_string("/home/axlefublr/.cache/mine/anki-card.html").expect("read card file");
    let mut lines = contents.lines().map(str::trim);

    let deck = lines.next().unwrap();
    if !VALID_DECKS.contains(&deck) {
        Err("invalid deck")?;
    }
    let note_type = lines.next().unwrap();
    if !VALID_NOTE_TYPES.contains(&note_type) {
        Err("invalid note type")?;
    }

    let mut lines: Vec<_> = lines.collect();
    lines.insert(0, note_type);
    lines.insert(0, deck);

    let fields = lines.len();
    if !(4..=6).contains(&fields) {
        Err(format!("expected 4..=6 fields, got {}", fields))?;
    }

    lines.resize(6, "");

    let card = lines
        .into_iter()
        .map(|line| format!("\"{}\"", line.replace('"', "\"\"")))
        .collect::<Vec<_>>()
        .join("|");
    println!("{}", card);
    Ok(())
}
